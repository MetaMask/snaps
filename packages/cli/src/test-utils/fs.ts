import { promises as fs } from 'fs';

jest.mock('fs');

export const BASE_PATH = '/cli';

/**
 * Clears out all the files in the in-memory file system, and writes the default
 * files to the `BASE_PATH` folder, including sub-folders.
 */
export async function resetFileSystem() {
  await fs.rm(BASE_PATH, { recursive: true, force: true });
}
