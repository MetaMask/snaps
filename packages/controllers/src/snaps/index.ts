export * from './utils';
export * from './SnapController';
export * from './endowments';
export * from './selectors';
