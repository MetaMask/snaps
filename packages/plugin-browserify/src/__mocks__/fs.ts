import { Volume } from 'memfs';

// Note: `Volume` implements most of the `fs` API, but not all.
const volume = new Volume();
export = { ...jest.requireActual('fs'), promises: volume.promises };
