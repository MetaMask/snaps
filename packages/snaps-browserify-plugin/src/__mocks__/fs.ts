import { Volume } from 'memfs';

// Note: `Volume` implements most of the `fs` API, but not all.
const volume = new Volume();

// Part of the real `fs` API is used here, since Browserify uses it to load
// certain modules from the file system. The async methods are not used, so
// they are mocked.
export = { ...jest.requireActual('fs'), promises: volume.promises };
