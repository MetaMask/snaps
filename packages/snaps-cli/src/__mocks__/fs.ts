import { Volume } from 'memfs';

// Note: `Volume` implements most of the `fs` API, but not all.
export default new Volume();
