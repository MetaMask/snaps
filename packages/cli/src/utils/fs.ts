import { promises as fs } from 'fs';
import type { Json } from '@metamask/snap-controllers';
import { logError } from './misc';

/**
 * Checks whether the given path string resolves to an existing directory, and
 * optionally creates the directory if it doesn't exist.
 *
 * @param pathString - The path string to check.
 * @param createDir - Whether to create the directory if it doesn't exist.
 * @returns Whether the given path is an existing directory.
 */
export async function isDirectory(
  pathString: string,
  createDir: boolean,
): Promise<boolean> {
  try {
    const stats = await fs.stat(pathString);
    return stats.isDirectory();
  } catch (error) {
    if (error.code === 'ENOENT') {
      if (!createDir) {
        return false;
      }

      try {
        await fs.mkdir(pathString);
        return true;
      } catch (mkdirError) {
        logError(`Directory '${pathString}' could not be created.`, mkdirError);
        throw mkdirError;
      }
    }
    return false;
  }
}

/**
 * Checks whether the given path string resolves to an existing file.
 *
 * @param pathString - The path string to check.
 * @returns Whether the given path is an existing file.
 */
export async function isFile(pathString: string): Promise<boolean> {
  try {
    const stats = await fs.stat(pathString);
    return stats.isFile();
  } catch (error) {
    return false;
  }
}

/**
 * Reads a `.json` file, parses its contents, and returns them.
 *
 * @param pathString - The path to the JSON file.
 * @returns The parsed contents of the JSON file.
 */
export async function readJsonFile(pathString: string): Promise<Json> {
  if (!pathString.endsWith('.json')) {
    throw new Error('The specified file must be a ".json" file.');
  }

  return JSON.parse(await fs.readFile(pathString, 'utf8'));
}
