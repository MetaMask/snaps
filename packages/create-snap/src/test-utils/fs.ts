import { promises as fs } from 'fs';

jest.mock('fs');

/**
 * Clears out all the files in the in-memory file system, and writes the default
 * folders to mimic the on-disk current working directory, including sub-folders.
 */
export async function resetFileSystem() {
  await fs.rm(process.cwd(), { recursive: true, force: true });

  // Using relative path with memfs won't work since process.cwd() specifies the cwd of the on-disk file system
  // This recursively creates the current working directory in the memfs volume.
  // Ref. https://github.com/streamich/memfs/blob/master/docs/relative-paths.md
  await fs.mkdir(process.cwd(), { recursive: true });
}
