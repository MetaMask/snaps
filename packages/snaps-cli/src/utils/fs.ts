import { promises as fs } from 'fs';
import { logError } from './misc';

/**
 * Checks whether the given path string resolves to an existing directory, and
 * optionally creates the directory if it doesn't exist.
 *
 * @param pathString - The path string to check
 * @param createDir - Whether to create the directory if it doesn't exist
 * @returns - Whether the given path is an existing directory
 */
export async function isDirectory(
  pathString: string,
  createDir: boolean,
): Promise<boolean> {
  try {
    const stats = await fs.stat(pathString);
    return stats.isDirectory();
  } catch (error) {
    if (error.code === 'ENOENT') {
      if (!createDir) {
        return false;
      }

      try {
        await fs.mkdir(pathString);
        return true;
      } catch (mkdirError) {
        logError(`Directory '${pathString}' could not be created.`, mkdirError);
        process.exit(1);
      }
    }
    return false;
  }
}

/**
 * Checks whether the given path string resolves to an existing file.
 *
 * @param {string} pathString - The path string to check
 * @returns {boolean} - Whether the given path is an existing file
 */
export async function isFile(pathString: string): Promise<boolean> {
  try {
    const stats = await fs.stat(pathString);
    return stats.isFile();
  } catch (error) {
    return false;
  }
}
