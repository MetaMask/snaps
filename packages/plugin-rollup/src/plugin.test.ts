import { promises as fs } from 'fs';
import virtual, { RollupVirtualOptions } from '@rollup/plugin-virtual';
import { OutputOptions, rollup, RollupOutput } from 'rollup';
import {
  DEFAULT_SNAP_BUNDLE,
  getSnapManifest,
} from '@metamask/snap-utils/test-utils';
import { checkManifest, evalBundle } from '@metamask/snap-utils';
import snaps, { Options } from './plugin';

jest.mock('fs');

jest.mock('@metamask/snap-utils', () => ({
  ...jest.requireActual('@metamask/snap-utils'),
  evalBundle: jest.fn(),
  checkManifest: jest.fn(),
}));

type BundleOptions = {
  code?: string;
  options?: Options;
  files?: RollupVirtualOptions;
  outputOptions?: OutputOptions;
  writeOptions?: OutputOptions;
  input?: string;
};

const bundle = async ({
  code = DEFAULT_SNAP_BUNDLE,
  options = { eval: false, manifestPath: undefined },
  files,
  outputOptions = {},
  writeOptions = { file: 'bundle.js' },
  input = 'foo',
}: BundleOptions = {}): Promise<RollupOutput> => {
  const bundler = await rollup({
    input,
    plugins: [
      virtual({
        foo: code,
        ...files,
      }),
      snaps(options),
    ],
    onwarn: (warning) => console.log(warning.message),
  });

  const output = await bundler.generate(outputOptions);
  await bundler.write(writeOptions);

  await bundler.close();

  return output;
};

describe('snaps', () => {
  it('processes files using Rollup', async () => {
    const { output } = await bundle();

    const { code } = output[0];
    expect(code).toMatchInlineSnapshot(`
      "module.exports.onRpcRequest = ({
        request
      }) => {
        console.log("Hello, world!");
        const {
          method,
          id
        } = request;
        return method + id;
      };
      "
    `);
  });

  it('applies a transform', async () => {
    const { output } = await bundle({
      code: `
        // foo bar
        /* baz qux */
        const foo = 'bar';
        console.log(foo);
      `,
    });

    const { code } = output[0];
    expect(code).toMatchInlineSnapshot(`
      "const foo = 'bar';
      console.log(foo);
      "
    `);
  });

  it('forwards the options', async () => {
    const { output } = await bundle({
      code: `
        // foo bar
        /* baz qux */
        const foo = 'bar';
        console.log(foo);
      `,
      options: {
        eval: false,
        manifestPath: undefined,
        stripComments: false,
      },
    });

    const { code } = output[0];
    expect(code).toMatchInlineSnapshot(`
      "// foo bar

      /* baz qux */
      const foo = 'bar';
      console.log(foo);
      "
    `);
  });

  it('runs on the entire bundle', async () => {
    const { output } = await bundle({
      code: `
        import { bar } from 'bar';

        // Sets foo to bar
        const foo = bar;
        console.log(foo);
      `,
      files: {
        bar: `
          // Returns baz
          export const bar = 'baz';
        `,
      },
    });

    const { code } = output[0];
    expect(code).toMatchInlineSnapshot(`
      "const bar = 'baz';
      const foo = bar;
      console.log(foo);
      "
    `);
  });

  it('generates a source map', async () => {
    await fs.writeFile('/source-map.ts', DEFAULT_SNAP_BUNDLE);

    const { output } = await bundle({
      // Rollup doesn't generate source maps from virtual files for some reason,
      // so we need to use a real file.
      input: '/source-map.ts',
      outputOptions: {
        sourcemap: true,
      },
    });

    const { map } = output[0];
    expect(map).toMatchInlineSnapshot(`
      SourceMap {
        "file": "source-map.js",
        "mappings": "AACEA,MAAM,CAACC,OAAP,CAAeC,YAAf,GAA8B,CAAC;EAAEC;AAAF,CAAD,KAAiB;EAC7CC,OAAO,CAACC,GAAR,CAAY,eAAZ;EAEA,MAAM;IAAEC,MAAF;IAAUC;EAAV,IAAiBJ,OAAvB;EACA,OAAOG,MAAM,GAAGC,EAAhB;AACD,CALD",
        "names": [
          "module",
          "exports",
          "onRpcRequest",
          "request",
          "console",
          "log",
          "method",
          "id",
        ],
        "sources": [
          "../../../../../../../source-map.ts",
        ],
        "sourcesContent": [
          "
        module.exports.onRpcRequest = ({ request }) => {
          console.log("Hello, world!");

          const { method, id } = request;
          return method + id;
        };
      ",
        ],
        "version": 3,
      }
    `);
  });

  it('evals the bundle if configured', async () => {
    const mock = evalBundle as jest.MockedFunction<typeof evalBundle>;
    mock.mockResolvedValue(null);

    await bundle({
      options: {
        eval: true,
        manifestPath: undefined,
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('bundle.js');
  });

  it('checks the manifest if configured', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: [],
    });

    await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('/', true, expect.any(String));
  });

  it('does not fix the manifest if configured', async () => {
    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: [],
    });

    await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
        writeManifest: false,
      },
    });

    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith('/', false, expect.any(String));
  });

  it('logs manifest errors if writeManifest is disabled', async () => {
    jest.spyOn(console, 'error').mockImplementation(() => undefined);

    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: ['foo', 'bar'],
      warnings: [],
    });

    await expect(
      bundle({
        options: {
          eval: false,
          manifestPath: '/snap.manifest.json',
          writeManifest: false,
        },
      }),
    ).rejects.toThrow('Manifest Error: The manifest is invalid.\nfoo\nbar');
  });

  it('logs manifest warnings', async () => {
    jest.spyOn(console, 'log').mockImplementation(() => undefined);

    const mock = checkManifest as jest.MockedFunction<typeof checkManifest>;
    mock.mockResolvedValue({
      manifest: getSnapManifest(),
      errors: [],
      warnings: ['foo', 'bar'],
    });

    await bundle({
      options: {
        eval: false,
        manifestPath: '/snap.manifest.json',
        writeManifest: false,
      },
    });

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith(
      'Manifest Warning: Validation of snap.manifest.json completed with warnings.\nfoo\nbar',
    );
  });

  it('forwards errors', async () => {
    const mock = evalBundle as jest.MockedFunction<typeof evalBundle>;
    mock.mockRejectedValue(new Error('foo'));

    await expect(
      bundle({
        options: {
          eval: true,
          manifestPath: undefined,
        },
      }),
    ).rejects.toThrow('foo');
  });

  it('shows a warning if no output file is configured', async () => {
    jest.spyOn(console, 'log').mockImplementation(() => undefined);

    await bundle({
      writeOptions: {
        dir: 'dist',
      },
    });

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(console.log).toHaveBeenCalledWith(
      'No output file specified, skipping bundle validation.',
    );
  });
});
