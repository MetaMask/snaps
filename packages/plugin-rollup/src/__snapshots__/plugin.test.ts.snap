// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snaps applies a transform 1`] = `
"
                        const foo = 'bar';
            console.log(foo);
"
`;

exports[`snaps forwards the options 1`] = `
"// foo bar
            /* baz qux */
            const foo = 'bar';
            console.log(foo);
"
`;

exports[`snaps processes files using Rollup 1`] = `
"const foo = 'bar';
            console.log(foo);
"
`;

exports[`snaps runs on the entire bundle 1`] = `
"
            const bar = 'baz';


            const foo = bar;
            console.log(foo);
"
`;
